### PR TITLE
Add new crypto features

### LINKED ISSUE #582: New crypto features support
### Issue To Be Solved

- add new features to the [crypto contract]( https://github.com/onflow/cadence/blob/master/runtime/stdlib/contracts/crypto.cdc), as required by the upcoming Flow protocol epoch features:
   - validation of public keys
   - support the signature scheme BLS and its corresponding hashing algorithm.
   
### Suggest A Solution

This is modified version of the file `crypto.cdc`. Many functions are copied from the original code but I pasted the entire file for a better clarity, as there a few changes here and there. I've added some comments in the code to help understand the changes. 

This code is a starting point to discuss and improve, I'm open to all suggestions making the APIs easier to use while keeping all the features needed for the epoch switch implementation in Flow protocol.

```
// the functions outside the crypto contract are meant to be implemented by the fvm 
// in Go. I can implement/support implementing these functions.

// the first 2 functions could be part of a same interface 
//(as they are both related to signature algos)

// verify is a low level signature verification function,
// that uses a signature scheme to verify a signature under hash(tag, data) and a public key.
pub fun verify(
    signature: [UInt8],
    tag: String,
    signedData: [UInt8],
    publicKey: [UInt8],
    signatureAlgorithm: String,
    hashAlgorithm: String
): Bool

// validatePublicKey is a low level function that checks if the input array 
// is a valid serialization of a public key of the given signature scheme.
pub fun validatePublicKey(
    publicKey: [UInt8],
    algorithm: String
): Bool

// low level hash function that computes hash(tag, data). 
// tag is mixed with data in a way dependent of the hash algo, this can be detailed in a doc later. 
// for instance, sha2(tag, data) = sha2(tag || data)
pub fun hash(
        data: [UInt8],
        tag: String,
        algorithm: String
    ): [UInt8] 


pub contract Crypto {

    // SignatureAlgorithm 
    pub struct SignatureAlgorithm {
        pub let name: String
        init(name: String) {       }
        
        // a method to validate a public key serialization
        pub fun validatePublicKey(publicKey: [UInt8]): Bool {
            return validatePublicKey(publicKey: publicKey, algorithm:self.name)
        }
        
        // a method to verify a signature
        pub fun verify(        
            signature: [UInt8],
            tag: String,
            signedData: [UInt8],
            publicKey: [UInt8],
            hashAlgorithm: HashAlgorithm ): Bool {

            return verify( 
                signature: signature, 
                tag: tag, 
                signedData: signedData, 
                publicKey: publicKey, 
                signatureAlgorithm:self.name, 
                hashAlgorithm: hashAlgorithm.name) 
        }

    }
    
   // supported SignatureAlgorithm
    
    // ECDSA  
    pub let ECDSA_P256: SignatureAlgorithm
    pub let ECDSA_Secp256k1: SignatureAlgorithm
    // BLS
    // why is it called BLS_BLS12381? 
    //   - BLS is the signature scheme, just like ECDSA
    //   - bls12381 is the name of the curve, just like P256 or Secp256k1
    //   - B, L and S in the scheme and the curve are not referring to the same people, and hence the redundancy :)  
    pub let BLS_BLS12381: SignatureAlgorithm 
    // with BLS being declared here, users could verify BLS signatures, only using the hasher KMAC_128_BLS_BLS12381 though (the Go implementation will enforce it).
    // But most importantly :
    //     - the staking contract can verify BLS public keys are valid.
    //     - the staking contract can verify a BLS proof of possession, required in the future 
    // for the BFT support


    // HashAlgorithm 
    pub struct HashAlgorithm {
        pub let name: String 
        init(name: String) {     }

        // tag is required for KMAC. 
        // sha2/sha3(tag, data) called with tag="" is equivalent to calling the standard sha2/sha3(data)
        pub fun hash(_ data: [UInt8], tag: String): [UInt8] {
            return hash(data: data, tag: tag, algorithm: self.name) 
        }
    }

    // supported HashAlgorithm 
    
    pub let SHA2_256: HashAlgorithm
    pub let SHA3_256: HashAlgorithm
    // KMAC_128_BLS_BLS12381 is an instance of KMAC_128 that this customized to be used with BLS and the curve BLS12381
    // why not declaring the generic KMAC_128? KMAC requires more customization inputs compared to a standard hash function, 
    // so it can't be supported by "HashAlgorithm". I already added "tag" to support the customized KMAC.
    pub let KMAC_128_BLS_BLS12381: HashAlgorithm 
    // could be added in another contract if considered too specific, the other contract would then also contain
    // the signature algo BLS_BLS12381


    // PublicKey is a structure with bytes and a signature algorithm
    pub struct PublicKey {
        pub let publicKey: [UInt8]
        pub let signatureAlgorithm: SignatureAlgorithm
        priv let valid: Bool
        // we could add: priv let valicationChecked: Bool to cash the validation result

        init(publicKey: [UInt8], signatureAlgorithm: SignatureAlgorithm) {
            self.publicKey = publicKey
            self.signatureAlgorithm = signatureAlgorithm
            self.valid = false
        }
        
        // not sure this getter in needed
        pub fun GetValidity(): Bool {
            return self.valid
        }

        // this is needed to verify ECDSA or BLS public keys are valid, 
        // which is required for the staking contract, but is also useful for other cases (there are many scenarios
        // where validating public keys is a must)
        pub fun Validate(): Bool {
            // we could check validationChecked here, to avoid re-validating
            self.valid = self.signatureAlgorithm.validatePublicKey(publicKey: self.publicKey)
            return self.valid
        }

        pub fun isValid(
            signature: [UInt8],
            signedData: [UInt8],
            domainSeparationTag: String,
            hashAlgorithm: HashAlgorithm
        ): Bool {
            // if we use valicationChecked, we can call self.Validate() here
            // to make sure self.valid is updated, without re-computating the result

            // Ensure the key is valid
            if !self.valid {
                return false
            }

            // Ensure the signature is valid
            return self.signatureAlgorithm.verify(
                signature: signature,
                tag: domainSeparationTag,
                signedData: signedData,
                publicKey: self.publicKey,
                hashAlgorithm: hashAlgorithm
            )
        }
    }

    // KeyListEntry is a structure representing one public key with other data
    pub struct KeyListEntry {
        pub let keyIndex: Int
        pub let publicKey: PublicKey
        pub let hashAlgorithm: HashAlgorithm
        pub let weight: UFix64
        pub let isRevoked: Bool

        init(
            keyIndex: Int,
            publicKey: PublicKey,
            hashAlgorithm: HashAlgorithm,
            weight: UFix64,
            isRevoked: Bool
        ) {
            self.keyIndex = keyIndex
            self.publicKey = publicKey
            self.hashAlgorithm = hashAlgorithm
            self.weight = weight
            self.isRevoked = isRevoked
        }
    }

    // KeyList is a list of KeyListEntry
    pub struct KeyList {
        priv let entries: [KeyListEntry]

        init() {
            self.entries = []
        }

        /// Adds a new key with the given weight
        pub fun add(
            _ publicKey: PublicKey,
            hashAlgorithm: HashAlgorithm,
            weight: UFix64
        ): KeyListEntry {

            let keyIndex = self.entries.length
            let entry = KeyListEntry(
                keyIndex: keyIndex,
                publicKey: publicKey,
                hashAlgorithm: hashAlgorithm,
                weight: weight,
                isRevoked: false
            )
            self.entries.append(entry)
            return entry
        }

        /// Returns the key at the given index, if it exists.
        /// Revoked keys are always returned, but they have `isRevoked` field set to true
        pub fun get(keyIndex: Int): KeyListEntry? {
            if keyIndex >= self.entries.length {
                return nil
            }

            return self.entries[keyIndex]
        }

        /// Marks the key at the given index revoked, but does not delete it
        pub fun revoke(keyIndex: Int) {
            if keyIndex >= self.entries.length {
                return
            }
            let currentEntry = self.entries[keyIndex]
            self.entries[keyIndex] = KeyListEntry(
                keyIndex: currentEntry.keyIndex,
                publicKey: currentEntry.publicKey,
                hashAlgorithm: currentEntry.hashAlgorithm,
                weight: currentEntry.weight,
                isRevoked: true
            )
        }

        /// Returns true if the given signatures are valid for the given signed data
        pub fun isValid(
            signatureSet: [KeyListSignature],
            signedData: [UInt8]
        ): Bool {
            var validWeights: UFix64 = 0.0
            let seenKeyIndices: {Int: Bool} = {}

            for signature in signatureSet {
                // Ensure the key index is valid
                if signature.keyIndex >= self.entries.length {
                    return false
                }

                // Ensure this key index has not already been seen
                if seenKeyIndices[signature.keyIndex] ?? false {
                    return false
                }

                // Record the key index was seen
                seenKeyIndices[signature.keyIndex] = true

                // Get the actual key
                let key = self.entries[signature.keyIndex]

                // Ensure the key is not revoked
                if key.isRevoked {
                    return false
                }

                // Ensure the signature is valid
                if !key.publicKey.isValid(
                    signature: signature.signature,
                    signedData: signedData,
                    domainSeparationTag: Crypto.domainSeparationTagUser,
                    hashAlgorithm: key.hashAlgorithm
                ) {
                    return false
                }

                validWeights = validWeights + key.weight
            }
            return validWeights >= 1.0
        }
    }

    // one signature entry
    pub struct KeyListSignature {
        pub let keyIndex: Int
        pub let signature: [UInt8]

        init(keyIndex: Int, signature: [UInt8]) {
            self.keyIndex = keyIndex
            self.signature = signature
        }
    }


    priv let domainSeparationTagUser: String

    init() {

        // Initialize constants

        self.ECDSA_P256 = SignatureAlgorithm(name: "ECDSA_P256")
        self.ECDSA_Secp256k1 = SignatureAlgorithm(name: "ECDSA_Secp256k1")
        self.BLS_BLS12381 = SignatureAlgorithm(name: "BLS_BLS12381")

        self.SHA2_256 = HashAlgorithm(name: "SHA2_256")
        self.SHA3_256 = HashAlgorithm(name: "SHA3_256")
        self.KMAC_128_BLS_BLS12381 = HashAlgorithm(name: "KMAC_128_BLS_BLS12381")

        self.domainSeparationTagUser = "user"
    }
}
```

### PR DESCRIPTION
**Pull request by SupunS**
Closes #582

## Description

This PR introduces the Cadence changes needed for the new crypto features. Includes the following changes:

#### Hash Algorithms
- New KMAC128_BLS_BLS12_381 hashing algorithm
  ```go
  var algo: HashAlgorithm = HashAlgorithm.KMAC128_BLS_BLS12_381
  ```

#### Signature Algorithms
- New BLS_BLS12_381 signature algorithm
  ```go
  var algo: SignatureAlgorithm = SignatureAlgorithm.BLS_BLS12_381
  ```
- Rename `ECDSA_Secp256k1` signature algorithm to `ECDSA_secp256k1`

#### PublicKey Native Type
- New `isValid` field
  - eg:
    ```go
    let publicKey = PublicKey(publicKey: [], signatureAlgorithm: SignatureAlgorithm.ECDSA_P256)

    let valid = publicKey.isValid
    ```
  - Get initialized during the construction of a public key, to the value returned by the `ValidatePublicKey` method on the host env.
- New `verify` method to validate a signature 
  - signature: `pub fun verify(signature: [UInt8], signedData: [UInt8], domainSeparationTag: String, hashAlgorithm: HashAlgorithm): Bool`
  - e.g:
    ```go
    let valid = publicKey.verify(
        signature: [],
        signedData: [],
        domainSeparationTag: "something",
        hashAlgorithm: HashAlgorithm.SHA2_256
    )
    ```
#### Crypto Contract
- New `hashWithTag` function
  - Signature: `pub fun hashWithTag(_ data: [UInt8], tag: string, algorithm: HashAlgorithm): [UInt8]` 
  - e.g:
    ```
    Crypto.hashWithTag(
        "01020304".decodeHex(),
        tag: "some-tag",
        algorithm: HashAlgorithm.SHA3_256
     )
    ``` 

### Changes to the FVM Interface
- New `ValidatePublicKey(key *PublicKey) (bool, error)` method
- Additional parameter ` tag string` in the `Hash` method.
  - New signature: `Hash(data []byte, tag string, hashAlgorithm HashAlgorithm) ([]byte, error)` 
- `PublicKey` has two new fields: `IsValid` and `Validated`
______

For contributor use:

- [x] Targeted PR against `master` branch
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [x] Code follows the [standards mentioned here](https://github.com/onflow/cadence/blob/master/CONTRIBUTING.md#styleguides)
- [x] Updated relevant documentation 
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Added appropriate labels

### CODE REVIEW

**Review by turbolent — APPROVED** [review]
Looks great, nice work! 

I added a breaking changes label because of the naming change from `ECDSA_Secp256k1` to `ECDSA_secp256k1`, which we need to communicate

File: docs/language/crypto.md [hunk 1]
```diff
@@ -99,13 +106,12 @@ The API of the Crypto contract is:
...
 
-    pub struct SignatureAlgorithm {
-        pub let name: String
-    }
-
     // Hash the data using the given hashing algorithm and returns the hashed data.
     pub fun hash(_ data: [UInt8], algorithm: HashAlgorithm): [UInt8]
 
+    // Hash the data using the given hashing algorithm and the tag. Returns the hashed data.
+    pub fun hashWithTag(_ data: [UInt8], tag: string, algorithm: HashAlgorithm): [UInt8]
```
Comment (tarakby) [inline docs/language/crypto.md:113]: I have suggested `hashWithTag` as I wasn't sure if `PublicKey.Verify` would be implemented entirely by the fvm or would be partly implemented in Cadence code. In the case of the former, I don't think `hashWithTag` is needed in Cadence, we would only need it as an fvm function.

File: docs/language/accounts.mdx [hunk 2]
```diff
@@ -167,6 +167,16 @@ struct AccountKey {
 struct PublicKey {
```
Comment (tarakby) [inline docs/language/accounts.mdx:167]: Can `PublicKey` fields (data, signature algo, isValid..) be updated after the creation of the object?

File: docs/language/accounts.mdx [hunk 3]
```diff
@@ -167,6 +167,16 @@ struct AccountKey {
...
     let publicKey: [UInt8]
     let signatureAlgorithm: SignatureAlgorithm
+    let isValid: Bool
+
+    /// Verifies a signature. Checks whether the signature was produced by signing
+    /// the given tag and data, using this public key and the given hash algorithm
+    pub fun verify(
```
Comment (tarakby) [inline docs/language/accounts.mdx:174]: As @turbolent mentioned about breaking changes, I think the change from `isValid` to `verify` is a breaking one.

File: docs/language/accounts.mdx (hunk 2 above)
Comment (SupunS) [reply docs/language/accounts.mdx:167]: No, they are defined as constant fields (i.e. read-only).

File: docs/language/accounts.mdx [hunk 4]
```diff
@@ -167,6 +167,16 @@ struct AccountKey {
 struct PublicKey {
     let publicKey: [UInt8]
     let signatureAlgorithm: SignatureAlgorithm
+    let isValid: Bool
+
+    /// Verifies a signature. Checks whether the signature was produced by signing
+    /// the given tag and data, using this public key and the given hash algorithm
```
[the remark this answers is not in the document]
Comment (SupunS) [reply docs/language/accounts.mdx:173]: I think it should be fine to add it here, or we can also add it as a note outside of the code block.

File: docs/language/accounts.mdx (hunk 3 above)
Comment (turbolent) [reply docs/language/accounts.mdx:174]: Hmm, both were added, so this isn't a breaking change

Comment (tarakby) [reply docs/language/accounts.mdx:174]: ah my bad, I somehow though `isValid` already existed.

**Pull request merged**